src/Form/EncryptionProfileForm.php

```diff
@@ -99,42 +83,66 @@ public function form(array $form, FormStateInterface $form_state) {
       '#disabled' => !$encryption_profile->isNew(),
     );
 
-    /** @var $key \Drupal\key\Entity\KeyInterface */
-    foreach ($keys as $key) {
-      $key_id = $key->id();
-      $key_title = $key->label();
-      $keys[$key_id] = (string) $key_title;
-    }
-
-    if ($profile_key = $encryption_profile->getEncryptionKey()) {
-      $default_key = $profile_key;
-    }
-
-    $form['encryption_key'] = array(
-      '#type' => 'select',
-      '#title' => $this->t('Encryption Key'),
-      '#description' => $this->t('Select the key used for encryption.'),
-      '#options' => $keys,
-      '#default_value' => (empty($default_key)?NULL:$default_key),
-      '#required' => TRUE
+    // This is the element that contains all of the dynamic parts of the form.
+    $form['encryption'] = array(
+      '#type' => 'container',
+      '#prefix' => '<div id="encrypt-settings">',
+      '#suffix' => '</div>',
     );
 
-    $enc_methods = [];
-    foreach ($this->encrypt_service->loadEncryptionMethods() as $plugin_id => $definition) {
-      $enc_methods[$plugin_id] = (string) $definition['title'];
+    $encryption_methods = $this->encryptService->loadEncryptionMethods();
+    $method_options = [];
+    $key_types = [];
+    foreach ($encryption_methods as $plugin_id => $definition) {
+      $method_options[$plugin_id] = (string) $definition['title'];
+      $key_types[$plugin_id] = $definition['key_types'];
     }
-    $form['encryption_method'] = array(
+
+    $form['encryption']['encryption_method'] = array(
       '#type' => 'select',
       '#title' => $this->t('Encryption Method'),
       '#description' => $this->t('Select the method used for encryption'),
-      '#options' => $enc_methods,
+      '#options' => $method_options,
       '#default_value' => $encryption_profile->getEncryptionMethod(),
+      '#ajax' => array(
+        'callback' => [$this, 'ajaxUpdateSettings'],
+        'event' => 'change',
+        'wrapper' => 'encrypt-settings',
+      ),
+    );
+
+    $current_encryption_method = $encryption_profile->getEncryptionMethod();
+    if (!$current_encryption_method && !empty($encryption_methods)) {
+      // Get the first one from the list.
+      $current_encryption_method = array_shift(array_keys($encryption_methods));
+    }
+
+    $form['encryption']['encryption_key'] = array(
+      '#type' => 'key_select',
+      '#title' => $this->t('Encryption Key'),
+      '#required' => TRUE,
+      '#key_filters' => ['type' => $key_types[$current_encryption_method]],
```

Comment:
I'm not familiar with this functionality. How does this list of keys get generated? I like the style, because we should be able to check if key_size is in the array of allowed sizes.